### PR TITLE
Attention Entropy

### PR DESCRIPTION
**Pull request by jaydeepborkar**
- Attention entropy and gini coefficient computation in `save_inference_log()`

### CODE REVIEW

File: inference.py [hunk 1]
```diff
@@ -198,9 +200,27 @@ def run_model_inferences(split_name: str, run_id: str, dataset: str, features: l
...
 
 
+def gini(array):
+    """Calculate the Gini coefficient of a numpy array. Ref: https://github.com/oliviaguest/gini"""
+    # based on bottom eq: http://www.statsdirect.com/help/content/image/stat0206_wmf.gif
+    # from: http://www.statsdirect.com/help/default.htm#nonparametric_methods/gini.htm
+    array = array.flatten() #all values are treated equally, arrays must be 1d
+    if np.amin(array) < 0:
+        array -= np.amin(array)  
+    #array += 0.0000001
```
Comment (Kyle1668) [inline inference.py:215]: Remove commented out code

File: inference.py [hunk 2]
```diff
@@ -198,9 +200,27 @@ def run_model_inferences(split_name: str, run_id: str, dataset: str, features: l
...
+    # from: http://www.statsdirect.com/help/default.htm#nonparametric_methods/gini.htm
+    array = array.flatten() #all values are treated equally, arrays must be 1d
+    if np.amin(array) < 0:
+        array -= np.amin(array)  
+    #array += 0.0000001  
+    array = np.sort(array)  
+    index = np.arange(1,array.shape[0]+1)  
+    n = array.shape[0] 
+    return ((np.sum((2 * index - n  - 1) * array)) / (n * np.sum(array)))  
+
```
Comment (Kyle1668) [inline inference.py:220]: Convention is two blank spaces between functions

File: inference.py [hunk 3]
```diff
@@ -229,9 +252,32 @@ def save_inference_log(
             inference_log["generation_perplexity"] = perplexities[index][1]
             inference_log["sequence_perplexity"] = perplexities[index][2]
         if "attn" in features:
-            for layer_index, attention_layer in enumerate(outputs.attentions):
-                sequence_attention = attention_layer[index].detach().tolist()
-                inference_log[f"attn_{layer_index}"] = sequence_attention
+            for layer_index, attention_layer in enumerate(outputs.attentions):
```
Comment (Kyle1668) [inline inference.py:255]: Let's separate this into it's own function somehow.

File: inference.py [hunk 4]
```diff
@@ -229,9 +252,32 @@ def save_inference_log(
...
         if "attn" in features:
-            for layer_index, attention_layer in enumerate(outputs.attentions):
-                sequence_attention = attention_layer[index].detach().tolist()
-                inference_log[f"attn_{layer_index}"] = sequence_attention
+            for layer_index, attention_layer in enumerate(outputs.attentions):  
+                sequence_attention = attention_layer[index].detach()  
+                head_e = []
+                gini_head = []
+
+                for head_index, head in enumerate(sequence_attention):
```
Comment (Kyle1668) [inline inference.py:260]: I don't know enough about this method to double-check the formula. I'd recommend getting a second look from Naomi or Katherine.

**Review by Kyle1668 — CHANGES_REQUESTED** [review]

File: inference.py (hunk 3 above)
Comment (jaydeepborkar) [reply inference.py:255]: I thought a bit about this. We are computing metrics for each attention head and saving them for each example. Having access to `index` and `inference_log` makes it easy. Do you have thoughts on how we can go about this?

File: inference.py (hunk 4 above)
Comment (jaydeepborkar) [reply inference.py:260]: Sure. Do you mean the logic behind calculating the gini coefficient, correct?

**Pull request closed without merging**